adaptdl/adaptdl/torch/adascale.py

```diff
@@ -227,26 +226,44 @@ def _final_callback(self):
             self._async_op.wait()
 
         grads = []
+        if self._mp_scaler is not None:
+            mixed_precision_scale = self._mp_scaler.get_scale()
+        else:
+            mixed_precision_scale = 1.0
         for group in self._optimizer.param_groups:
             grads.append([])
             for param in group["params"]:
                 if param.grad is None:
                     grads[-1].append(None)
                     continue
-                param.grad.div_(self._accum_count)
-                grads[-1].append(param.grad.detach())
-
-        check = [g.sum() for group in grads for g in group]
-        if any(c != c or c in (float('inf'), -float('inf')) for c in check):
+                grad = param.grad.detach().float()
+                grads[-1].append(
+                    grad / mixed_precision_scale / self._accum_count)
+
+        # Note: mixed precision can result in nan/inf gradients,
+        # which propogate into our norm and variance estimates.
+        # Mixed precision autoscaling skips the skip where
+        # there are nan/inf, so we also skip the update here
+        grads_normsqr = _normsqr_groups(grads)
+        if not np.all(np.isfinite(grads_normsqr)):
             print("AdaScale detected invalid gradient! Skipping step.")
+            # Note: failing to zero here will accumulate the gradients,
+            # which will infinitely propogate the nan's/infs
+            # Also note: zeroing here will cause automatic mixed precision to
+            # never back off
+            self._accum_count = 0
```

Comment:
Note: I'd much rather have `_reset_accumulation()` not call `self._orig_optimizer_zero_grad()`, and then call `_reset_accumulation()` here.

Reply:
I think you can just call `_reset_accumulation` instead of `_orig_optimizer_zero_grad` in `zero_grad`. No need to reset any variables here, and that eliminates any chance that the gradients, `local_sqr`, and `accum_count` becomes de-synchronized due to programmer error.